### PR TITLE
feat(SSC): package-lock.json v3 support

### PR DESCRIPTION
**Pull request by mmcqd**
We got it

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: cli/src/semdep/parsers/package_lock.py [hunk 1]
```diff
@@ -22,6 +23,96 @@
...
+                transitivity=Transitivity(Transitive())
+                if nested
+                else transitivity(manifest_deps, [package]),
+                line_number=dep_json.line_number,
+            )
+        )
+    return output
+
+
+def parse_dependencies_field(
```
Comment (mmcqd) [inline cli/src/semdep/parsers/package_lock.py:67]: This got moved and renamed, but not re-written

File: cli/src/semdep/parsers/package_lock.py [hunk 2]
```diff
@@ -22,6 +23,96 @@
...
 
+def parse_packages_field(deps: Dict[str, JSON]) -> List[FoundDependency]:
+    deps_json = deps.get("packages")
+    if deps_json is None:
+        logger.debug("Found package-lock with no 'packages'")
+        return []
+
+    deps = deps_json.as_dict()
+    try:
+        manifest_deps = set(deps[""].as_dict()["dependencies"].as_dict().keys())
```
Comment (adamberman) [inline cli/src/semdep/parsers/package_lock.py:28]: the value you're looking for here typically has an empty string key?

File: cli/src/semdep/parsers/package_lock.py [hunk 3]
```diff
@@ -22,6 +23,96 @@
 logger = getLogger(__name__)
 
 
+def parse_packages_field(deps: Dict[str, JSON]) -> List[FoundDependency]:
+    deps_json = deps.get("packages")
+    if deps_json is None:
+        logger.debug("Found package-lock with no 'packages'")
+        return []
+
+    deps = deps_json.as_dict()
```
Comment (adamberman) [inline cli/src/semdep/parsers/package_lock.py]: you're overwriting an existing variable, one that comes from the arguments. this is (i think) typically not best practice in python

File: cli/src/semdep/parsers/package_lock.py [hunk 4]
```diff
@@ -22,6 +23,96 @@
...
+    try:
+        manifest_deps = set(deps[""].as_dict()["dependencies"].as_dict().keys())
+    except KeyError:
+        manifest_deps = None
+    output = []
+    for package, dep_json in deps.items():
+        fields = dep_json.as_dict()
+        version = fields.get("version")
+        package_name = package[
+            package.rfind("node_modules") + 12 :
```
Comment (adamberman) [inline cli/src/semdep/parsers/package_lock.py]: ... + 12?

File: cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarejs-yarn2-sca.yaml-dependency_awarepackage-lock-v3/results.txt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+=== stdout - plain
+{
+  "errors": [],
+  "paths": {
+    "_comment": "<add --verbose for a list of skipped paths>",
+    "scanned": [
+      "targets/dependency_aware/package-lock-v3/package-lock.json"
+    ]
+  },
+  "results": [],
```
Comment (adamberman) [inline cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarejs-yarn2-sca.yaml-dependency_awarepackage-lock-v3/results.txt]: this says no results, so this likely doesn't work yet?

File: cli/src/semdep/parsers/package_lock.py (hunk 4 above)
Comment (mmcqd) [reply cli/src/semdep/parsers/package_lock.py]: The length of the string "node_modules". Basically we want to extract `foo` from something like `node_modules/bar/node_modules/foo`

File: cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarejs-yarn2-sca.yaml-dependency_awarepackage-lock-v3/results.txt (hunk 5 above)
Comment (mmcqd) [reply cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarejs-yarn2-sca.yaml-dependency_awarepackage-lock-v3/results.txt]: well... you would be right.

File: cli/src/semdep/parsers/package_lock.py (hunk 2 above)
Comment (mmcqd) [reply cli/src/semdep/parsers/package_lock.py:28]: yes, check the example package-lock.json

File: cli/src/semdep/parsers/package_lock.py (hunk 4 above)
Comment (mmcqd) [reply cli/src/semdep/parsers/package_lock.py]: should've been 13 lol

**Review by brendongo — APPROVED** [review]

**Pull request merged**